File: Server/src/main/core/game/node/item/ChanceItem.java
package core.game.node.item;

import core.game.node.entity.npc.drop.DropFrequency;
import core.tools.RandomFunction;

/**
 * Chance item.
 */
public final class ChanceItem extends Item {

    /**
     * The constant DROP_RATES.
     */
    public static final int[] DROP_RATES = {5, 15, 150, 750};

    private double chanceRate;

    private int minimumAmount;

    private int maximumAmount;

    private int tableSlot;

    private DropFrequency dropFrequency;

    private int setRate = -1;

    /**
     * Instantiates a new Chance item.
     *
     * @param id the id
     */
    public ChanceItem(int id) {
        this(id, 1, 1, 1000, 1.0);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param chanceRate    the chance rate
     */
    public ChanceItem(int id, int minimumAmount, double chanceRate) {
        this(id, minimumAmount, minimumAmount, 1000, chanceRate);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param chanceRate    the chance rate
     */
    public ChanceItem(int id, int minimumAmount, int chanceRate) {
        this(id, minimumAmount, minimumAmount, 1000, (double) chanceRate);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param maximumAmount the maximum amount
     * @param chanceRate    the chance rate
     */
    public ChanceItem(int id, int minimumAmount, int maximumAmount, double chanceRate) {
        this(id, minimumAmount, maximumAmount, 1000, chanceRate);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param maximumAmount the maximum amount
     * @param frequency     the frequency
     */
    public ChanceItem(int id, int minimumAmount, int maximumAmount, DropFrequency frequency) {
        this(id, minimumAmount, maximumAmount, 1000, DropFrequency.rate(frequency));
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param maximumAmount the maximum amount
     * @param charge        the charge
     * @param chanceRate    the chance rate
     */
    public ChanceItem(int id, int minimumAmount, int maximumAmount, int charge, double chanceRate) {
        this(id, minimumAmount, maximumAmount, charge, chanceRate, DropFrequency.UNCOMMON);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param maximumAmount the maximum amount
     * @param charge        the charge
     * @param chanceRate    the chance rate
     * @param frequency     the frequency
     */
    public ChanceItem(int id, int minimumAmount, int maximumAmount, int charge, double chanceRate, DropFrequency frequency) {
        this(id, minimumAmount, maximumAmount, charge, chanceRate, frequency, -1);
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id            the id
     * @param minimumAmount the minimum amount
     * @param maximumAmount the maximum amount
     * @param charge        the charge
     * @param chanceRate    the chance rate
     * @param frequency     the frequency
     * @param setRate       the set rate
     */
    public ChanceItem(int id, int minimumAmount, int maximumAmount, int charge, double chanceRate, DropFrequency frequency, int setRate) {
        super(id, minimumAmount, charge);
        this.minimumAmount = minimumAmount;
        this.maximumAmount = maximumAmount;
        this.chanceRate = chanceRate;
        this.dropFrequency = frequency;
        this.setRate = setRate;
    }

    /**
     * Instantiates a new Chance item.
     *
     * @param id        the id
     * @param amount    the amount
     * @param frequency the frequency
     */
    public ChanceItem(int id, int amount, DropFrequency frequency) {
        this(id, amount, amount, frequency);
    }

    /**
     * Gets random item.
     *
     * @return the random item
     */
    public Item getRandomItem() {
        if (minimumAmount == maximumAmount) {
            return new Item(getId(), minimumAmount);
        }
        return new Item(getId(), RandomFunction.random(minimumAmount, maximumAmount));
    }

    /**
     * Gets item.
     *
     * @param table the table
     * @return the item
     */
    public static ChanceItem getItem(ChanceItem... table) {
        return getItem(RandomFunction.getRandomDouble(75.0), table);
    }

    /**
     * Gets item.
     *
     * @param chance the chance
     * @param table  the table
     * @return the item
     */
    public static ChanceItem getItem(double chance, ChanceItem... table) {
        // TODO:
        return table[0];
    }

    /**
     * Gets copy.
     *
     * @return the copy
     */
    public ChanceItem getCopy() {
        ChanceItem newItem = new ChanceItem(getId(), minimumAmount, maximumAmount, getCharge(), chanceRate, dropFrequency);
        return newItem;
    }

    /**
     * Gets chance rate.
     *
     * @return the chance rate
     */
    public double getChanceRate() {
        return chanceRate;
    }

    /**
     * Sets chance rate.
     *
     * @param chanceRate the chance rate
     */
    public void setChanceRate(double chanceRate) {
        this.chanceRate = chanceRate;
    }

    /**
     * Gets minimum amount.
     *
     * @return the minimum amount
     */
    public int getMinimumAmount() {
        return minimumAmount;
    }

    /**
     * Sets minimum amount.
     *
     * @param minimumAmount the minimum amount
     */
    public void setMinimumAmount(int minimumAmount) {
        this.minimumAmount = minimumAmount;
    }

    /**
     * Gets maximum amount.
     *
     * @return the maximum amount
     */
    public int getMaximumAmount() {
        return maximumAmount;
    }

    /**
     * Sets maximum amount.
     *
     * @param maximumAmount the maximum amount
     */
    public void setMaximumAmount(int maximumAmount) {
        this.maximumAmount = maximumAmount;
    }

    /**
     * Gets drop frequency.
     *
     * @return the drop frequency
     */
    public DropFrequency getDropFrequency() {
        return dropFrequency;
    }

    /**
     * Sets drop frequency.
     *
     * @param dropFrequency the drop frequency
     */
    public void setDropFrequency(DropFrequency dropFrequency) {
        this.dropFrequency = dropFrequency;
    }

    /**
     * Gets table slot.
     *
     * @return the table slot
     */
    public int getTableSlot() {
        return tableSlot;
    }

    /**
     * Sets table slot.
     *
     * @param tableSlot the table slot
     */
    public void setTableSlot(int tableSlot) {
        this.tableSlot = tableSlot;
    }

    @Override
    public String toString() {
        return "ChanceItem " + super.toString() + " [chanceRate=" + chanceRate + ", minimumAmount=" + minimumAmount + ", maximumAmount=" + maximumAmount + ", tableSlot=" + tableSlot + ", dropFrequency=" + dropFrequency + "]";
    }

    /**
     * Gets set rate.
     *
     * @return the set rate
     */
    public int getSetRate() {
        return setRate;
    }

    /**
     * Sets set rate.
     *
     * @param setRate the set rate
     */
    public void setSetRate(int setRate) {
        this.setRate = setRate;
    }
}

File: Emulator/src/main/core/game/node/item/ChanceItem.kt
package core.game.node.item

import core.game.node.entity.npc.drop.DropFrequency
import core.tools.RandomFunction

/**
 * Represents an item with a chance-rate.
 * @author Emperor
 */
class ChanceItem @JvmOverloads constructor(
    id: Int,
    minimumAmount: Int = 1,
    maximumAmount: Int = 1,
    charge: Int = 1000,
    chanceRate: Double = 1.0,
    frequency: DropFrequency? = DropFrequency.UNCOMMON,
    setRate: Int = -1
) :
    Item(id, minimumAmount, charge) {
    /**
     * @return the chanceRate.
     */
    /**
     * @param chanceRate the chanceRate to set.
     */
    /**
     * The chance rate of this item.
     */
    var chanceRate: Double = 0.0

    /**
     * @return the minimumAmount.
     */
    /**
     * @param minimumAmount the minimumAmount to set.
     */
    /**
     * The minimum amount.
     */
    var minimumAmount: Int = 0

    /**
     * @return the maximumAmount.
     */
    /**
     * @param maximumAmount the maximumAmount to set.
     */
    /**
     * The maximum amount.
     */
    var maximumAmount: Int = 0

    /**
     * @return the tableSlot.
     */
    /**
     * @param tableSlot the tableSlot to set.
     */
    /**
     * The table slot.
     */
    var tableSlot: Int = 0

    /**
     * @return the dropFrequency.
     */
    /**
     * @param dropFrequency the dropFrequency to set.
     */
    /**
     * The drop frequency.
     */
    var dropFrequency: DropFrequency? = null

    /**
     * Gets the setRate.
     * @return the setRate
     */
    /**
     * Sets the basetRate.
     * @param setRate the setRate to set.
     */
    /**
     * The set rate.
     */
    var setRate: Int = -1

    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     * @param minimumAmount The minimum amount.
     * @param chanceRate The chance rate.
     */
    constructor(id: Int, minimumAmount: Int, chanceRate: Double) : this(
        id,
        minimumAmount,
        minimumAmount,
        1000,
        chanceRate
    )

    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     * @param minimumAmount The minimum amount.
     * @param maximumAmount The maximum amount.
     * @param chanceRate The chance rate.
     */
    constructor(id: Int, minimumAmount: Int, maximumAmount: Int, chanceRate: Double) : this(
        id,
        minimumAmount,
        maximumAmount,
        1000,
        chanceRate
    )

    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id the id.
     * @param minimumAmount the min amount.
     * @param maximumAmount the max amount.
     * @param frequency the frequency.
     */
    constructor(id: Int, minimumAmount: Int, maximumAmount: Int, frequency: DropFrequency?) : this(
        id,
        minimumAmount,
        maximumAmount,
        1000,
        DropFrequency.rate(frequency).toDouble()
    )

    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     * @param minimumAmount The minimum amount.
     * @param maximumAmount The maximum amount.
     * @param charge The charge.
     * @param frequency The drop frequency.
     * @param chanceRate The chance rate.
     * @param setRate the set rate.
     */
    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     * @param minimumAmount The minimum amount.
     * @param maximumAmount The maximum amount.
     * @param charge The charge.
     * @param frequency The drop frequency.
     * @param chanceRate The chance rate.
     */
    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     * @param minimumAmount The minimum amount.
     * @param maximumAmount The maximum amount.
     * @param charge The charge.		if(frequency == DropFrequency.ALWAYS){
     *
     * }
     * @param chanceRate The chance rate.
     */
    /**
     * Constructs a new `ChanceItem` `Object`.
     * @param id The item id.
     */
    init {
        this.minimumAmount = minimumAmount
        this.maximumAmount = maximumAmount
        this.chanceRate = chanceRate
        this.dropFrequency = frequency
        this.setRate = setRate
    }

    constructor(id: Int, amount: Int, frequency: DropFrequency?) : this(id, amount, amount, frequency)

    val randomItem: Item
        /**
         * Gets the item instance.
         * @return the item.
         */
        get() {
            if (minimumAmount == maximumAmount) {
                return Item(id, minimumAmount)
            }
            return Item(id, RandomFunction.random(minimumAmount, maximumAmount))
        }

    val copy: ChanceItem
        /**
         * Gets a copy.
         * @return the copy.
         */
        get() {
            val newItem = ChanceItem(id, minimumAmount, maximumAmount, charge, chanceRate, dropFrequency)
            return newItem
        }

    override fun toString(): String {
        return "ChanceItem " + super.toString() + " [chanceRate=" + chanceRate + ", minimumAmount=" + minimumAmount + ", maximumAmount=" + maximumAmount + ", tableSlot=" + tableSlot + ", dropFrequency=" + dropFrequency + "]"
    }

    companion object {
        val DROP_RATES: IntArray = intArrayOf(5, 15, 150, 750)

        /**
         * Gets a random chance item from the table.
         * @param table The table.
         * @return The chance item.
         */
        fun getItem(vararg table: ChanceItem): ChanceItem {
            return getItem(RandomFunction.getRandomDouble(75.0), *table)
        }

        /**
         * Gets a random chance item from the table.
         * @param table The table.
         * @return The chance item.
         */
        fun getItem(chance: Double, vararg table: ChanceItem): ChanceItem {
            // TODO:
            return table[0]
        }
    }
}